### PR TITLE
Recompose

### PR DESCRIPTION
**Pull request by SaraIsabelSantacruz**
## Summary
- createTypes and completeTypes added
-actionsCreators changed like redux-recompose suggests
-createReducer, completeReducer and completeState added
-reducers changed like redux-recompose suggests
-fetchMiddleware added

## Trello Card
https://trello.com/c/jw1x90Qv

### CODE REVIEW

File: react/tic-tac-toe-api/index.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const jsonServer = require('json-server');
 const server = jsonServer.create();
-const router = jsonServer.router('db.json');
+const router = jsonServer.router('db.json');
```
Comment (gon-escalante) [inline react/tic-tac-toe-api/index.js]: remove this space

File: react/tic-tac-toe/src/redux/data/actions.js [hunk 2]
```diff
@@ -1,40 +1,21 @@
...
-  GET_DATA_REQUEST: '@@DATA/GET_DATA_REQUEST',
-  GET_DATA_SUCCESS: '@@DATA/GET_DATA_SUCCESS',
-  GET_DATA_FAILURE: '@@DATA/GET_DATA_FAILURE',
-  CREATE_DATA_REQUEST: '@@CREATE_DATA_REQUEST',
-  CREATE_DATA_SUCCESS: '@@CREATE_DATA_SUCCESS',
-  CREATE_DATA_FAILURE: '@@CREATE_DATA_FAILURE'
-};
+import FetchService from '../../services/MatchesService';
+
+export const actions = createTypes(completeTypes(['GET_DATA', 'CREATE_DATA', 'ON_LOGIN', 'SET_AUTH']), '@@SOCCER');
```
Comment (gon-escalante) [inline react/tic-tac-toe/src/redux/data/actions.js]: `ON_LOGIN` and `SET_AUTH` dont belong here

File: react/tic-tac-toe/src/redux/data/reducers.js [hunk 3]
```diff
@@ -1,49 +1,16 @@
...
-        loading: false,
-        error: true
-      };
-    default:
-      return state;
-  }
+const initialState = completeState(stateDescription);
+
+const game = {
+  primaryActions: [actions.GET_DATA, actions.CREATE_DATA, actions.ON_LOGIN]
```
Comment (gon-escalante) [inline react/tic-tac-toe/src/redux/data/reducers.js]: `ON_LOGIN` doesnt belong here

File: react/tic-tac-toe/src/redux/login/actions.js [hunk 4]
```diff
@@ -1,35 +1,33 @@
...
+
 import dataFetch from '../../services/MatchesService';
 
-export const actions = {
-  SET_AUTH: '@@DATA/SET_AUTH',
-  LOGIN_REQUEST: '@@DATA/LOGIN_REQUEST',
-  LOGIN_SUCCESS: '@@DATA/GLOGIN_SUCCESS',
-  LOGIN_FAILURE: '@@DATA/LOGIN_FAILURE'
-};
+export const actions = createTypes(completeTypes(['ON_LOGIN', 'SET_AUTH', 'ON_LOGOUT']), '@@SOCCER');
```
Comment (gon-escalante) [inline react/tic-tac-toe/src/redux/login/actions.js]: i dont think `SET_AUTH` and `ON_LOGOUT` need the complete types, remember that `completeTypes` creates the actions `SET_AUTH_FAILURE` and `SET_AUTH_SUCCESS` to handle the loading state, and set auth and loading dont seem to need a loading state.

File: react/tic-tac-toe/src/redux/login/actions.js [hunk 5]
```diff
@@ -1,35 +1,33 @@
...
-    dispatch({ type: actions.LOGIN_SUCCESS, hasToken, isError: false });
-    return true;
-  },
-  logOut: () => dispatch => {
-    localStorage.removeItem('token');
-    dispatch({ type: actions.LOGIN_FAILURE, hasToken: false, isError: false });
-  }
+  onLogin: values => ({
+    type: actions.ON_LOGIN,
+    target: 'setToken',
```
Comment (CrisVega08) [inline react/tic-tac-toe/src/redux/login/actions.js]: Use a constant for targets

File: react/tic-tac-toe/src/redux/login/actions.js [hunk 6]
```diff
@@ -1,35 +1,33 @@
...
+        dispatch({
+          type: actions.SET_AUTH,
+          target: 'isUserLoggedin'
+        });
+      })
+    ]
+  }),
+  setAuth: () => ({
+    type: actions.SET_AUTH,
+    target: 'isUserLoggedin'
```
Comment (CrisVega08) [inline react/tic-tac-toe/src/redux/login/actions.js]: Same

File: react/tic-tac-toe/src/redux/login/actions.js [hunk 7]
```diff
@@ -1,35 +1,33 @@
...
+  }),
+  logOut: () => ({
+    type: actions.ON_LOGOUT,
+    target: 'isUserLoggedin'
```
Comment (CrisVega08) [inline react/tic-tac-toe/src/redux/login/actions.js]: Same

**Review by CrisVega08 — CHANGES_REQUESTED** [review]
Great work with redux-recompose

**Review by gon-escalante — APPROVED** [review]
nice work 👏

**Pull request merged**